DigitalLearningSolutions.Data.Migrations/202204071530_PopulateUsersTableFromAccountsTables.cs

```diff
@@ -0,0 +1,415 @@
+                    GETUTCDATE(),
+                    GETUTCDATE()
+                    FROM DelegateAccounts
+                    WHERE Email IN (
+                        SELECT Email FROM DelegateAccounts
+                        WHERE Email IS NOT NULL AND TRIM(Email) IS NOT NULL AND TRIM(Email) <> ''
+                        GROUP BY Email
+                        HAVING COUNT(*) = 1
+                        EXCEPT
+                        SELECT PrimaryEmail FROM Users)"
+            );
+            connection.Execute(
+                @"UPDATE DelegateAccounts
+                    SET
+                        UserID = (SELECT ID FROM Users WHERE Email = PrimaryEmail),
+                        Email = NULL,
+                        CentreSpecificDetailsLastChecked = GETUTCDATE()"
+            );
+
+            // Get the rest of the delegate accounts we've not resolved yet
+            var delegateAccounts =
+                connection.Query<DelegateAccount>("SELECT * FROM DelegateAccounts WHERE UserId IS NULL");
+
+            var delegateAccountsGroupedByEmail = delegateAccounts.GroupBy(da => da.Email);
+
+            foreach (var emailGroup in delegateAccountsGroupedByEmail)
+            {
+                if (!string.IsNullOrWhiteSpace(emailGroup.Key))
+                {
+                    var delegatesToAttemptToLink = new List<DelegateAccount>();
+                    var delegateAccountsGroupedByCentre = emailGroup.GroupBy(da => da.CentreId);
+
+                    foreach (var centreGroup in delegateAccountsGroupedByCentre)
+                    {
+                        delegatesToAttemptToLink.Add(centreGroup.First());
+
+                        // All duplicate emails at a centre need new User accounts.
+                        // If there are more than one delegate accounts with the same email at a centre
+                        // we only want to keep the email on the first. The rest get reset to having a User with a guid email
+                        var othersAtCentre = centreGroup.Skip(1);
+
+                        foreach (var delegateAccount in othersAtCentre)
+                        {
+                            // Insert a new User with Guid email
+                            InsertNewUserForDelegateAccount(connection, delegateAccount, true);
+                        }
+                    }
+
+                    var allJobGroupsMatch =
+                        delegatesToAttemptToLink.Select(da => da.JobGroupId_deprecated).Distinct().Count() < 2;
+                    foreach (var delegateAccount in delegatesToAttemptToLink)
+                    {
+                        // Check for existing user with email
+                        var existingUserWithEmail = connection.QuerySingleOrDefault<User>(
+                            "SELECT * FROM Users WHERE PrimaryEmail = @email",
+                            new { email = emailGroup.Key }
+                        );

```

Comment:
I think this `existingUserWithEmail` definition can be moved outside of this forEach. Saves repeated calling.

UPDATE: realised we are using the repeated calling to check the updated details and compare them in an iterative methodology. I feel like this can all be done in a single call, where we do a lot of checks on the delegatesToAttemptToLink prior to it. I'd prefer that to repeatedly updating the same record multiple times, but not sure how feasible it is.

Reply:
I don't really see how we can do this when the account will not exist for the first delegate with the email, but not for later ones. The migration doesn't really take very long to run, and given it only needs to run once (assuming all goes well) I don't think it's worth spending forever optimising absolutely everything.

Reply:
Fair enough. For what it's worth, I was envisioning a process where we check all the accounts first, take whatever details we would need to enter to build the "User" class, and then insert it (or update it if it exists). But then the logic to determine what to place in there would be more pernickety. What we have already implemented here is a lot simpler, I just didn't get what was going on until I'd read through it all.